adapter/service/virtualmeter/virtual_meter_manager.go

```diff
@@ -0,0 +1,332 @@
+	vmeter, ok := meter.(*virtualMeterManager)
+	if !ok {
+		log.Fatal("failed to inject adapter into virtual meter")
+	}
+
+	vmeter.ad = ad
+
+	return vmeter
+}
+
+// WithTaskManager adds a provided task manager to the provided virtual meter manager. Used to avoid circular dependencies.
+func WithTaskManager(meter Manager, manager task.Manager) Manager {
+	vmeter, ok := meter.(*virtualMeterManager)
+	if !ok {
+		log.Fatal("failed to inject adapter into virtual meter")
+	}
+
+	vmeter.taskManager = manager
+
+	return vmeter
+}
+
+func (m *virtualMeterManager) Add(addr string, modes map[string]float64, unit string) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	addr = m.thingAddrFromTopic(addr)
+
+	s := m.serviceTemplates[addr]
+	if s == nil {
+		return fmt.Errorf("failed to add meter to the thing: %s. no service template found. %v", addr, m.serviceTemplates)
+	}
+
+	thing := m.ad.ThingByAddress(addr)
+	if thing == nil {
+		return fmt.Errorf("no thing found by address: %s. can't add meter", addr)
+	}
+
+	device, err := m.storage.Device(addr)
+	if err != nil {
+		return fmt.Errorf("failed to get device: %w", err)
+	}
+
+	oldModes := device.Modes
+	if oldModes != nil {
+		if err := m.recalculateEnergy(&device); err != nil {
+			return fmt.Errorf("failed to update energy: %w", err)
+		}
+	}
+
+	device.Modes = modes
+	device.Unit = unit
+
+	if err := m.storage.SetDeviceEntry(addr, device); err != nil {
+		return fmt.Errorf("failed add meter, can't save data: %w", err)
+	}
+
+	// Update thing only if a service has been just added.
+	if oldModes == nil {
+		if err := thing.Update(true, adapter.ThingUpdateAddService(s)); err != nil {
+			return fmt.Errorf("failed to update thing. Can't add service. Addr: %s. %w", addr, err)
+		}
+	}
+
+	return nil
+}
+
+func (m *virtualMeterManager) Remove(addr string) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	addr = m.thingAddrFromTopic(addr)
+
+	s := m.serviceTemplates[addr]
+	if s == nil {
+		return fmt.Errorf("failed to remove meter from a thing: %s. No service template found", addr)
+	}
+
+	thing := m.ad.ThingByAddress(addr)
+	if thing == nil {
+		return fmt.Errorf("no thing found by address: %s. can't remove meter", addr)
+	}
+
+	if err := m.storage.DeleteDevice(addr); err != nil {
+		return fmt.Errorf("failed to delete meter, can't remove from storage: %w", err)
+	}
+
+	if err := thing.Update(true, adapter.ThingUpdateRemoveService(m.serviceTemplates[addr])); err != nil {
+		return fmt.Errorf("failed to update thing. Can't remove service. Addr: %s. %w", addr, err)
+	}
+
+	return nil
+}
+
+func (m *virtualMeterManager) Modes(addr string) (map[string]float64, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+
+	addr = m.thingAddrFromTopic(addr)
+
+	device, err := m.storage.Device(addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get modes: %w", err)
+	}
+
+	return device.Modes, nil
+}
+
+func (m *virtualMeterManager) SetReportingInterval(duration time.Duration) error {
```

Comment:
We totally don't need this in this implementation I believe.

Interval at which meter values will be reported is purely dependent on the interval set in numerimeter.Task(). This task will periodically poll for reports all meter elec services it finds.